lib/api/BackbeatAPI.js

```diff
@@ -0,0 +1,208 @@
+                    if (obj.metadata && obj.metadata[this._repConfig.topic]) {
+                        const copy = JSON.parse(JSON.stringify(obj.metadata[
+                            this._repConfig.topic]));
+                        topicMD.metadata = copy;
+                        response.splice(idx, 1);
+                    }
+                });
+                response.push(topicMD);
+
+                connections.isrHealth = this._checkISRHealth(topicMD.metadata);
+            } finally {
+                Object.assign(connections, this._getConnectionDetails());
+
+                response.push({
+                    internalConnections: connections,
+                });
+
+                return cb(null, response);
+            }
+        });
+    }
+
+    setupInternals(cb) {
+        async.parallel([
+            done => this._setZookeeper(done),
+            done => this._setProducer(this._metricsTopic, (err, producer) => {
+                if (err) {
+                    return done(err);
+                }
+                this._metricProducer = producer;
+                return done();
+            }),
+            done => this._setProducer(this._crrTopic, (err, producer) => {
+                if (err) {
+                    return done(err);
+                }
+                this._crrProducer = producer;
+                return done();
+            }),
+            done => this._setProducer(this._crrStatusTopic, (err, producer) => {
+                if (err) {
+                    return done(err);
+                }
+                this._crrStatusProducer = producer;
+                return done();
+            }),
+        ], err => {
+            if (err) {
+                this._logger.error('error setting up internal clients');
+                return cb(err);
+            }
+            this._logger.info('BackbeatAPI setup ready');
+            return cb();
+        });
+    }
+
+    _setProducer(topic, cb) {
+        const producer = new BackbeatProducer({
+            zookeeper: { connectionString: this._zkConfig.connectionString },
+            topic,
+        });
+
+        producer.once('error', cb);
+        producer.once('ready', () => {
+            producer.removeAllListeners('error');
+            producer.on('error', err => {
+                this._logger.error('error from backbeat producer', {
+                    error: err,
+                });
+                return cb(err);
+            });
+            return cb(null, producer);
+        });
+    }
+
+    _setZookeeper(cb) {
+        const populatorZkPath = this._queuePopulator.zookeeperPath;
+        const zookeeperUrl =
+            `${this._zkConfig.connectionString}${populatorZkPath}`;
+
+        const zkClient = zookeeper.createClient(zookeeperUrl, {
+            autoCreateNamespace: this._zkConfig.autoCreateNamespace,
+        });
+        zkClient.connect();
+
+        zkClient.once('error', cb);
+        zkClient.once('state', event => {
```

Comment:
You can also use the `'connected'` event, so you don't have to check the events (and you may receive other kinds of events that you want to ignore rather than trigger an error if not what you're waiting for).